In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle
%matplotlib inline

In [1]:
from google.colab import files
uploaded = files.upload()

Saving train.csv to train.csv


In [3]:
df_train = pd.read_csv('train.csv')
df_train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  127         120       RL          NaN     4928   Pave   NaN      IR1   
1  889          20       RL         95.0    15865   Pave   NaN      IR1   
2  793          60       RL         92.0     9920   Pave   NaN      IR1   
3  110          20       RL        105.0    11751   Pave   NaN      IR1   
4  422          20       RL          NaN    16635   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
4         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2007        WD         Normal     128000  
1     10   2007        WD         Normal     268000  
2      6   2007        WD         Normal     269790  
3      1   2010       COD         Normal     190000  
4      6   2009        WD         Normal     215000  

[5 rows x 81 columns]

In [4]:
df_train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1168.000000  1168.000000    954.00000    1168.000000  1168.000000   
mean    724.136130    56.767979     70.98847   10484.749144     6.104452   
std     416.159877    41.940650     24.82875    8957.442311     1.390153   
min       1.000000    20.000000     21.00000    1300.000000     1.000000   
25%     360.500000    20.000000     60.00000    7621.500000     5.000000   
50%     714.500000    50.000000     70.00000    9522.500000     6.000000   
75%    1079.500000    70.000000     80.00000   11515.500000     7.000000   
max    1460.000000   190.000000    313.00000  164660.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1168.000000  1168.000000   1168.000000  1161.000000  1168.000000  ...   
mean      5.595890  1970.930651   1984.758562   102.310078   444.726027  ...   
std       1.124343    30.145255     20.785185   182.595606   462.664785  ...   
min       1.000000  1875.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1966.000000     0.000000     0.000000  ...   
50%       5.000000  1972.000000   1993.000000     0.000000   385.500000  ...   
75%       6.000000  2000.000000   2004.000000   160.000000   714.500000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1168.000000  1168.000000    1168.000000  1168.000000  1168.000000   
mean     96.206336    46.559932      23.015411     3.639555    15.051370   
std     126.158988    66.381023      63.191089    29.088867    55.080816   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    24.000000       0.000000     0.000000     0.000000   
75%     171.000000    70.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1168.000000   1168.000000  1168.000000  1168.000000    1168.000000  
mean      3.448630     47.315068     6.344178  2007.804795  181477.005993  
std      44.896939    543.264432     2.686352     1.329738   79105.586863  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  130375.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163995.000000  
75%       0.000000      0.000000     8.000000  2009.000000  215000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In this data set we have many categorical columns.
column lotfrontage is having 954 values it means it have null values which is need to be treated. 

In [6]:
from sklearn.impute import KNNImputer 
knnimp = KNNImputer(n_neighbors=2)
df_file = pd.DataFrame(knnimp.fit_transform(df_train[['OverallQual','LotFrontage']]))
df_file.columns=(['OverallQual','LotFrontage'])
df_file

OverallQual  LotFrontage
0             6.0         77.5
1             8.0         95.0
2             7.0         92.0
3             6.0        105.0
4             6.0         77.5
...           ...          ...
1163          5.0         79.0
1164          4.0         67.0
1165          6.0         24.0
1166          4.0         50.0
1167          6.0         77.5

[1168 rows x 2 columns]

In [7]:
df_train.drop(['OverallQual','LotFrontage'],axis = 1,inplace = True)

In [8]:
# seperating int type columns
df_int = df_train.select_dtypes(include='int')

In [9]:
df_Tr=pd.concat([df_file,df_int],axis = 1)

In [5]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1168 non-null   int64  
 1   MSSubClass     1168 non-null   int64  
 2   MSZoning       1168 non-null   object 
 3   LotFrontage    954 non-null    float64
 4   LotArea        1168 non-null   int64  
 5   Street         1168 non-null   object 
 6   Alley          77 non-null     object 
 7   LotShape       1168 non-null   object 
 8   LandContour    1168 non-null   object 
 9   Utilities      1168 non-null   object 
 10  LotConfig      1168 non-null   object 
 11  LandSlope      1168 non-null   object 
 12  Neighborhood   1168 non-null   object 
 13  Condition1     1168 non-null   object 
 14  Condition2     1168 non-null   object 
 15  BldgType       1168 non-null   object 
 16  HouseStyle     1168 non-null   object 
 17  OverallQual    1168 non-null   int64  
 18  OverallC

column number 6,57,72,73,74 have so many null values, so its better to drop such column

In [10]:
df_train.drop(['Alley','FireplaceQu','PoolQC','Fence','MiscFeature'],axis = 1,inplace = True)

In [11]:
# only int data type columns 
df_Tr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 36 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OverallQual    1168 non-null   float64
 1   LotFrontage    1168 non-null   float64
 2   Id             1168 non-null   int64  
 3   MSSubClass     1168 non-null   int64  
 4   LotArea        1168 non-null   int64  
 5   OverallCond    1168 non-null   int64  
 6   YearBuilt      1168 non-null   int64  
 7   YearRemodAdd   1168 non-null   int64  
 8   BsmtFinSF1     1168 non-null   int64  
 9   BsmtFinSF2     1168 non-null   int64  
 10  BsmtUnfSF      1168 non-null   int64  
 11  TotalBsmtSF    1168 non-null   int64  
 12  1stFlrSF       1168 non-null   int64  
 13  2ndFlrSF       1168 non-null   int64  
 14  LowQualFinSF   1168 non-null   int64  
 15  GrLivArea      1168 non-null   int64  
 16  BsmtFullBath   1168 non-null   int64  
 17  BsmtHalfBath   1168 non-null   int64  
 18  FullBath

### all values are complete but column no 6,7 are needed to be converted as they are date type so need to convert them to number of years 

In [12]:
# seperating categorical data 
df_cat= df_train.select_dtypes(include='object')

In [13]:
df_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 38 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSZoning       1168 non-null   object
 1   Street         1168 non-null   object
 2   LotShape       1168 non-null   object
 3   LandContour    1168 non-null   object
 4   Utilities      1168 non-null   object
 5   LotConfig      1168 non-null   object
 6   LandSlope      1168 non-null   object
 7   Neighborhood   1168 non-null   object
 8   Condition1     1168 non-null   object
 9   Condition2     1168 non-null   object
 10  BldgType       1168 non-null   object
 11  HouseStyle     1168 non-null   object
 12  RoofStyle      1168 non-null   object
 13  RoofMatl       1168 non-null   object
 14  Exterior1st    1168 non-null   object
 15  Exterior2nd    1168 non-null   object
 16  MasVnrType     1161 non-null   object
 17  ExterQual      1168 non-null   object
 18  ExterCond      1168 non-null

we need to convert the all he object data type column based on their values so below is the coversion of all the columns based on the type of value they have , if it is ordinal type we have used here ordinal encoder, if it is a simple category we have used label encoder, if the number of category is more then 6 we have used binary encoder .

In [14]:
!pip install category_encoders

In [15]:
from category_encoders import BinaryEncoder
bin_enc = BinaryEncoder(cols = ['MSZoning'])
a = bin_enc.fit_transform(df_cat['MSZoning'])


In [16]:
from sklearn.preprocessing import LabelEncoder
lab_enc = LabelEncoder()
Lot= pd.DataFrame(lab_enc.fit_transform(df_cat['Street']))
Lot.columns=(['Street'])


In [17]:
Lots= pd.DataFrame(lab_enc.fit_transform(df_cat['LotShape']))
Lots.columns=(['LotShape'])


In [18]:
Lotc= pd.DataFrame(lab_enc.fit_transform(df_cat['LotConfig']))
Lotc.columns=(['LotConfig'])


In [19]:
LotS= pd.DataFrame(lab_enc.fit_transform(df_cat['LandSlope']))
LotS.columns=(['LandSlope'])


In [20]:

from category_encoders import BinaryEncoder
bin_enc = BinaryEncoder(cols = ['Neighborhood'])
nd = bin_enc.fit_transform(df_cat['Neighborhood'])


In [21]:

bin_enc = BinaryEncoder(cols = ['Condition1'])
cd = bin_enc.fit_transform(df_cat['Condition1'])


In [22]:

bin_enc = BinaryEncoder(cols = ['Condition2'])
cd2 = bin_enc.fit_transform(df_cat['Condition2'])


In [23]:
Bld= pd.DataFrame(lab_enc.fit_transform(df_cat['BldgType']))
Bld.columns=(['BldgType'])


In [24]:
Hst= pd.DataFrame(lab_enc.fit_transform(df_cat['HouseStyle']))
Hst.columns=(['HouseStyle'])


In [25]:
rft= pd.DataFrame(lab_enc.fit_transform(df_cat['RoofStyle']))
rft.columns=(['RoofStyle'])


In [26]:
rftm= pd.DataFrame(lab_enc.fit_transform(df_cat['RoofMatl']))
rftm.columns=(['RoofMatl'])


In [27]:

bin_enc = BinaryEncoder(cols = ['Exterior1st'])
ext1 = bin_enc.fit_transform(df_cat['Exterior1st'])


In [28]:

bin_enc = BinaryEncoder(cols = ['Exterior2nd'])
ext2 = bin_enc.fit_transform(df_cat['Exterior2nd'])


In [29]:
Mas= pd.DataFrame(lab_enc.fit_transform(df_cat['MasVnrType']))
Mas.columns=(['MasVnrType'])


In [30]:
df_cat.drop(['MasVnrType','Exterior2nd','Exterior1st','RoofMatl','RoofStyle','HouseStyle','BldgType',
             'Condition2','Condition1','Neighborhood','LandSlope','LotConfig','LotShape','Street','MSZoning'],axis =1,inplace = True)

/usr/local/lib/python3.8/dist-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [31]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_cat['ExterQual'] = df_cat.ExterQual.map(dic)

<ipython-input-31-5a75090bf2a8>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['ExterQual'] = df_cat.ExterQual.map(dic)


In [33]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_cat['ExterCond'] = df_cat.ExterCond.map(dic)


<ipython-input-33-c61bb1a178f5>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['ExterCond'] = df_cat.ExterCond.map(dic)


In [34]:
fond= pd.DataFrame(lab_enc.fit_transform(df_cat['Foundation']))
fond.columns=(['Foundation'])


In [36]:
df_cat.drop(['Foundation'],axis = 1, inplace = True)

In [37]:

from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_cat['BsmtQual'] = df_cat.BsmtQual.map(dic)

<ipython-input-37-1aae0c0d5d82>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['BsmtQual'] = df_cat.BsmtQual.map(dic)


In [38]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_cat['BsmtCond'] = df_cat.BsmtCond.map(dic)

<ipython-input-38-11ccd16a46a9>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['BsmtCond'] = df_cat.BsmtCond.map(dic)


In [39]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'AllPub':4,'NoSewr':3,'NoSeWa':2,'ELO':1}
df_cat['Utilities'] = df_cat.Utilities.map(dic)

<ipython-input-39-054ec3955ec3>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['Utilities'] = df_cat.Utilities.map(dic)


In [40]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'Gd':5,'Av':4,'Mn':3,'No':2,'NA':1}
df_cat['BsmtExposure'] = df_cat.BsmtExposure.map(dic)

<ipython-input-40-671eab09ec9a>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['BsmtExposure'] = df_cat.BsmtExposure.map(dic)


In [41]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'GLQ':7,'ALQ':6,'BLQ':5,'Rec':4,'LwQ':3,'Unf':2,'NA':1}
df_cat['BsmtFinType1'] = df_cat.BsmtFinType1.map(dic)

<ipython-input-41-781d80a97c9d>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['BsmtFinType1'] = df_cat.BsmtFinType1.map(dic)


In [42]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'GLQ':7,'ALQ':6,'BLQ':5,'Rec':4,'LwQ':3,'Unf':2,'NA':1}
df_cat['BsmtFinType2'] = df_cat.BsmtFinType2.map(dic)

<ipython-input-42-2ed43ee717ab>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['BsmtFinType2'] = df_cat.BsmtFinType2.map(dic)


In [43]:
from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_cat['HeatingQC'] = df_cat.HeatingQC.map(dic)

<ipython-input-43-6c73c23d3353>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['HeatingQC'] = df_cat.HeatingQC.map(dic)


In [45]:
heat= pd.DataFrame(lab_enc.fit_transform(df_cat['Heating']))
heat.columns=(['Heating'])


In [46]:
centair= pd.DataFrame(lab_enc.fit_transform(df_cat['CentralAir']))
centair.columns=(['CentralAir'])


In [47]:
elc= pd.DataFrame(lab_enc.fit_transform(df_cat['Electrical']))
elc.columns=(['Electrical'])


In [48]:
df_cat.drop(['Heating','CentralAir','Electrical'],axis = 1,inplace = True)

/usr/local/lib/python3.8/dist-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [49]:

from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_cat['KitchenQual'] = df_cat.KitchenQual.map(dic)

<ipython-input-49-868ed162da87>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['KitchenQual'] = df_cat.KitchenQual.map(dic)


In [50]:

from sklearn.preprocessing import OrdinalEncoder
dic = {'Typ':8,'Min1':7,' Min2':6,'Mod':5,'Maj1':4,'Maj2':3,'Sev':2,'Sal':1}
df_cat['Functional'] = df_cat.Functional.map(dic)

<ipython-input-50-f89d80c9e7bc>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['Functional'] = df_cat.Functional.map(dic)


In [51]:

from sklearn.preprocessing import OrdinalEncoder
dic = {'2Types':7,'Attchd':6,'Basment':5,'BuiltIn':4,'CarPort':3,'Detchd':2,'NA':1}
df_cat['GarageType'] = df_cat.GarageType.map(dic)

<ipython-input-51-0da459ca0a3e>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['GarageType'] = df_cat.GarageType.map(dic)


In [52]:

from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_cat['GarageQual'] = df_cat.GarageQual.map(dic)


from sklearn.preprocessing import OrdinalEncoder
dic = {'Fin':4,'RFn':3,'Unf':2,'NA':1}
df_cat['GarageFinish'] = df_cat.GarageFinish.map(dic)


from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_cat['GarageCond'] = df_cat.GarageCond.map(dic)

Pav= pd.DataFrame(lab_enc.fit_transform(df_cat['PavedDrive']))
Pav.columns=(['PavedDrive'])


<ipython-input-52-73dd42353008>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['GarageQual'] = df_cat.GarageQual.map(dic)
<ipython-input-52-73dd42353008>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cat['GarageFinish'] = df_cat.GarageFinish.map(dic)
<ipython-input-52-73dd42353008>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.o

In [53]:
df_cat.drop(['PavedDrive'],axis = 1, inplace = True)

In [54]:
df_cat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   LandContour    1168 non-null   object 
 1   Utilities      1168 non-null   int64  
 2   ExterQual      1168 non-null   int64  
 3   ExterCond      1168 non-null   int64  
 4   BsmtQual       1138 non-null   float64
 5   BsmtCond       1138 non-null   float64
 6   BsmtExposure   1137 non-null   float64
 7   BsmtFinType1   1138 non-null   float64
 8   BsmtFinType2   1137 non-null   float64
 9   HeatingQC      1168 non-null   int64  
 10  KitchenQual    1168 non-null   int64  
 11  Functional     1138 non-null   float64
 12  GarageType     1104 non-null   float64
 13  GarageFinish   1104 non-null   float64
 14  GarageQual     1104 non-null   float64
 15  GarageCond     1104 non-null   float64
 16  SaleType       1168 non-null   object 
 17  SaleCondition  1168 non-null   object 
dtypes: float

In [55]:
L_C= pd.DataFrame(lab_enc.fit_transform(df_cat['LandContour']))
L_C.columns=(['LandContour'])

In [56]:

bin_enc = BinaryEncoder(cols = ['SaleType'])
S_L_T = bin_enc.fit_transform(df_cat['SaleType'])

In [57]:
bin_enc = BinaryEncoder(cols = ['SaleCondition'])
S_L_C = bin_enc.fit_transform(df_cat['SaleCondition'])

In [58]:
df_cat.drop(['LandContour','SaleType','SaleCondition'],axis = 1,inplace = True)

In [59]:
df_cat_final = pd.concat([df_cat,S_L_C,S_L_T,L_C,Pav,elc,centair,
                          heat,fond,Mas,ext2,ext1,rftm,rft,Hst,Bld,
                          cd2,cd,nd,LotS,Lotc,Lots,Lot,a],axis =1)

In [60]:
df_cat_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 61 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Utilities        1168 non-null   int64  
 1   ExterQual        1168 non-null   int64  
 2   ExterCond        1168 non-null   int64  
 3   BsmtQual         1138 non-null   float64
 4   BsmtCond         1138 non-null   float64
 5   BsmtExposure     1137 non-null   float64
 6   BsmtFinType1     1138 non-null   float64
 7   BsmtFinType2     1137 non-null   float64
 8   HeatingQC        1168 non-null   int64  
 9   KitchenQual      1168 non-null   int64  
 10  Functional       1138 non-null   float64
 11  GarageType       1104 non-null   float64
 12  GarageFinish     1104 non-null   float64
 13  GarageQual       1104 non-null   float64
 14  GarageCond       1104 non-null   float64
 15  SaleCondition_0  1168 non-null   int64  
 16  SaleCondition_1  1168 non-null   int64  
 17  SaleCondition_

In [61]:
df_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1168 entries, 0 to 1167
Data columns (total 34 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Id             1168 non-null   int64
 1   MSSubClass     1168 non-null   int64
 2   LotArea        1168 non-null   int64
 3   OverallCond    1168 non-null   int64
 4   YearBuilt      1168 non-null   int64
 5   YearRemodAdd   1168 non-null   int64
 6   BsmtFinSF1     1168 non-null   int64
 7   BsmtFinSF2     1168 non-null   int64
 8   BsmtUnfSF      1168 non-null   int64
 9   TotalBsmtSF    1168 non-null   int64
 10  1stFlrSF       1168 non-null   int64
 11  2ndFlrSF       1168 non-null   int64
 12  LowQualFinSF   1168 non-null   int64
 13  GrLivArea      1168 non-null   int64
 14  BsmtFullBath   1168 non-null   int64
 15  BsmtHalfBath   1168 non-null   int64
 16  FullBath       1168 non-null   int64
 17  HalfBath       1168 non-null   int64
 18  BedroomAbvGr   1168 non-null   int64
 19  Kitche

In [62]:

df_int['Current_year'] = 2022

<ipython-input-62-afae4ea725e9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_int['Current_year'] = 2022


In [63]:
df_int['Years_old']=df_int['Current_year']-df_int['YearBuilt']

<ipython-input-63-1d02f7e5054a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_int['Years_old']=df_int['Current_year']-df_int['YearBuilt']


In [64]:
df_int['remodel_duration'] = df_int['Current_year']-df_int['YearRemodAdd']

<ipython-input-64-cc105fbba698>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_int['remodel_duration'] = df_int['Current_year']-df_int['YearRemodAdd']


In [65]:
df_int.drop(['Current_year','YearBuilt','YearRemodAdd'],axis = 1, inplace = True)

In [ ]:
df_int.info()

column number 29 and 30 have month and year data which is needed to be treated 

In [67]:
df_int.MoSold.unique()

array([ 2, 10,  6,  1, 11,  5,  4,  7,  8,  3,  9, 12])

In [69]:
# they are actually month number but our M.L dosent understand such value it will concider it as a number only and it will treat them like 11>2 or 4<7 which is needed to be treated 
bin_enc = BinaryEncoder(cols = ['MoSold'])
mosold = bin_enc.fit_transform(df_int['MoSold'])

In [71]:
df_int.YrSold.unique()

array([2007, 2010, 2009, 2006, 2008])

In [72]:
bin_enc = BinaryEncoder(cols = ['YrSold'])
yrsold = bin_enc.fit_transform(df_int['YrSold'])

In [ ]:
df_int.drop(['YrSold','MoSold'],axis = 1,inplace = True)

In [75]:
df_Train_final= pd.concat([df_int,mosold,yrsold,df_cat_final],axis = 1)

In [ ]:
df_Train_final.info()

In [77]:
Data_Train = df_Train_final.dropna(axis = 0)

In [120]:
# trainning dataset is ready now with no null value and no object data type
Data_Train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1051 entries, 0 to 1167
Data columns (total 100 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                1051 non-null   int64  
 1   MSSubClass        1051 non-null   int64  
 2   LotArea           1051 non-null   int64  
 3   OverallCond       1051 non-null   int64  
 4   BsmtFinSF1        1051 non-null   int64  
 5   BsmtFinSF2        1051 non-null   int64  
 6   BsmtUnfSF         1051 non-null   int64  
 7   TotalBsmtSF       1051 non-null   int64  
 8   1stFlrSF          1051 non-null   int64  
 9   2ndFlrSF          1051 non-null   int64  
 10  LowQualFinSF      1051 non-null   int64  
 11  GrLivArea         1051 non-null   int64  
 12  BsmtFullBath      1051 non-null   int64  
 13  BsmtHalfBath      1051 non-null   int64  
 14  FullBath          1051 non-null   int64  
 15  HalfBath          1051 non-null   int64  
 16  BedroomAbvGr      1051 non-null   int64  

In [114]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler # for standardizing the Data
#import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
import pickle
from sklearn.metrics import r2_score

import warnings
warnings.filterwarnings

<function warnings.filterwarnings(action, message='', category=<class 'Warning'>, module='', lineno=0, append=False)>

In [121]:
for i in Data_Train:
  print(Data_Train[i].skew())

0.03334518858946745
1.3975490154229993
10.384923992652418
0.8739660112144471
1.8943665631071365
4.256117993357226
0.923583725345719
2.510279087680288
1.567545288631137
0.7823600960133135
9.88149118109127
1.5287394680901565
0.4748168130625055
4.082598226706046
0.0014348039584395115
0.5292587560526554
-0.11490868484546496
6.097354078038587
0.6379107715731476
0.6012505491237023
0.18116891205373964
0.8379046363949282
1.4832775465141421
2.242513059084789
3.1737022044095475
9.743446453136132
3.980871272380127
12.55274712810775
22.086264169442327
2.0174377576182154
0.6725111548884651
0.5776652303081885
0.3661405619028324
0.551467912274173
-0.36213869144490485
0.04765524510995009
0.2711507461198851
0.7050880725044089
-0.6869989954393471
0
0.7692208105508352
1.7909854268388465
0.23399082776709154
0.39499849759214883
1.1436506194228468
-0.21279553826871764
3.430441966306644
-0.5805597602455238
0.34934003359442417
-6.602085756813843
-0.7870115472320081
0.31300766461566065
-1.4297519251798514
-3.6

In [ ]:
for i in Data_Train:
  print("upper limit",Data_Train[i].mean()+3*Data_Train[i].std())
  print("Lower limit",Data_Train[i].mean()-3*Data_Train[i].std())


In [124]:
from scipy.stats import zscore

In [135]:
for i in Data_Train:
  Data_Train['Z_score'] = zscore(Data_Train[i]) 

<ipython-input-135-7b20b60642e2>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Data_Train['Z_score'] = zscore(Data_Train[i])


In [147]:
Data_Train.Z_score.head(20)

0     0.411189
1     0.411189
2     0.411189
3     0.411189
4     0.411189
5     0.411189
6     0.411189
7     0.411189
8     0.411189
9     0.411189
10   -2.431969
11    0.411189
12    0.411189
13    0.411189
14    0.411189
15    0.411189
16   -2.431969
17    0.411189
18    0.411189
19   -2.431969
Name: Z_score, dtype: float64

In [151]:
Data_Train[Data_Train['Z_score']<-3]

Empty DataFrame
Columns: [Id, MSSubClass, LotArea, OverallCond, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, TotRmsAbvGrd, Fireplaces, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, SalePrice, Years_old, remodel_duration, MoSold_0, MoSold_1, MoSold_2, MoSold_3, YrSold_0, YrSold_1, YrSold_2, Utilities, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, Functional, GarageType, GarageFinish, GarageQual, GarageCond, SaleCondition_0, SaleCondition_1, SaleCondition_2, SaleType_0, SaleType_1, SaleType_2, SaleType_3, LandContour, PavedDrive, Electrical, CentralAir, Heating, Foundation, MasVnrType, Exterior2nd_0, Exterior2nd_1, Exterior2nd_2, Exterior2nd_3, Exterior1st_0, Exterior1st_1, Exterior1st_2, Exterior1st_3, RoofMatl, RoofStyle, HouseStyle, BldgType, Condition2_0, Condition2_1, Condition2_2, Condition2_3, Condition1_0, Condition1_1, Condition1_2, Condition1_3, Neighborhood_0, Neighborhood_1, Neighborhood_2, Neighborhood_3, Neighborhood_4, LandSlope, LotConfig, LotShape, Street, MSZoning_0, MSZoning_1, MSZoning_2, ...]
Index: []

[0 rows x 101 columns]

NO outliers

In [153]:
Data_Train[Data_Train['Z_score']>3]

Empty DataFrame
Columns: [Id, MSSubClass, LotArea, OverallCond, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, TotRmsAbvGrd, Fireplaces, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, SalePrice, Years_old, remodel_duration, MoSold_0, MoSold_1, MoSold_2, MoSold_3, YrSold_0, YrSold_1, YrSold_2, Utilities, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, Functional, GarageType, GarageFinish, GarageQual, GarageCond, SaleCondition_0, SaleCondition_1, SaleCondition_2, SaleType_0, SaleType_1, SaleType_2, SaleType_3, LandContour, PavedDrive, Electrical, CentralAir, Heating, Foundation, MasVnrType, Exterior2nd_0, Exterior2nd_1, Exterior2nd_2, Exterior2nd_3, Exterior1st_0, Exterior1st_1, Exterior1st_2, Exterior1st_3, RoofMatl, RoofStyle, HouseStyle, BldgType, Condition2_0, Condition2_1, Condition2_2, Condition2_3, Condition1_0, Condition1_1, Condition1_2, Condition1_3, Neighborhood_0, Neighborhood_1, Neighborhood_2, Neighborhood_3, Neighborhood_4, LandSlope, LotConfig, LotShape, Street, MSZoning_0, MSZoning_1, MSZoning_2, ...]
Index: []

[0 rows x 101 columns]

NO outliers

In [163]:
# Now handling MultiCollinearity 
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif = pd.DataFrame()
vif['Vif'] = Data_Train.columns
vif["Vif"] = [variance_inflation_factor(Data_Train.values, i)
                          for i in range(len(Data_Train.columns))]
vif['Vif'].head(20)

0     1.102634
1     5.976469
2     1.915045
3     2.252006
4          inf
5          inf
6          inf
7          inf
8          inf
9          inf
10         inf
11         inf
12    2.562887
13    1.280418
14    3.835979
15    2.594627
16    2.831992
17    1.785807
18    5.374967
19    1.768525
Name: Vif, dtype: float64

In [164]:
Df_Tr = Data_Train.drop(Data_Train.columns[[4,5,6,7,8,9,10,11]],axis = 1)

In [166]:
Df_Tr.drop(['Z_score'],axis = 1,inplace = True)

In [170]:
a= pd.DataFrame()
a['vif'] = Df_Tr.columns

a["Vif"] = [variance_inflation_factor(Df_Tr.values, i)
                          for i in range(len(Df_Tr.columns))]


3125.549134467213

In [174]:
a[a['Vif']>9]

vif          Vif
31        Utilities  3125.549134
48  SaleCondition_2    10.130705
51       SaleType_2    10.530818

We have chacked the outliers using z_score method. we have checked the multicollienearity now we will do the same thing on test data set 

In [175]:
DF = Df_Tr.drop(Df_Tr.columns[[31,48,51]],axis = 1)

In [177]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1051 entries, 0 to 1167
Data columns (total 89 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                1051 non-null   int64  
 1   MSSubClass        1051 non-null   int64  
 2   LotArea           1051 non-null   int64  
 3   OverallCond       1051 non-null   int64  
 4   BsmtFullBath      1051 non-null   int64  
 5   BsmtHalfBath      1051 non-null   int64  
 6   FullBath          1051 non-null   int64  
 7   HalfBath          1051 non-null   int64  
 8   BedroomAbvGr      1051 non-null   int64  
 9   KitchenAbvGr      1051 non-null   int64  
 10  TotRmsAbvGrd      1051 non-null   int64  
 11  Fireplaces        1051 non-null   int64  
 12  GarageCars        1051 non-null   int64  
 13  GarageArea        1051 non-null   int64  
 14  WoodDeckSF        1051 non-null   int64  
 15  OpenPorchSF       1051 non-null   int64  
 16  EnclosedPorch     1051 non-null   int64  


In [93]:
# now we need to perform the all the activity on test data set and making it ready like our train data set .
from google.colab import files
uploaded = files.upload()

Saving test.csv to test (1).csv


In [ ]:
df_test = pd.read_csv('test.csv')
df_test.head()

In [ ]:
df_test.info()

In [ ]:
from sklearn.impute import KNNImputer 
knnimp = KNNImputer(n_neighbors=2)
df_f = pd.DataFrame(knnimp.fit_transform(df_test[['OverallQual','LotFrontage']]))
df_f.columns=(['OverallQual','LotFrontage'])
df_f

In [96]:
df_test.drop(['OverallQual','LotFrontage'],axis = 1,inplace = True)

In [ ]:
df_test.describe()

In [98]:
# seperating int type columns
df_int = df_test.select_dtypes(include='int')
# seperating obj type columns
df_obj = df_test.select_dtypes(include='object')

In [ ]:
df_int.info()

In [ ]:
# performing the same thing on column number 4,5,31,32
# they are actually month number but our M.L dosent understand such value it will concider it as a number only and it will treat them like 11>2 or 4<7 which is needed to be treated 
bin_enc = BinaryEncoder(cols = ['MoSold'])
mosold = bin_enc.fit_transform(df_int['MoSold'])

bin_enc = BinaryEncoder(cols = ['YrSold'])
yrsold = bin_enc.fit_transform(df_int['YrSold'])

df_int.drop(['MoSold','YrSold'],axis = 1, inplace = True)

df_int['Current_year'] = 2022
df_int['Years_old']=df_int['Current_year']-df_int['YearBuilt']
df_int['remodel_duration'] = df_int['Current_year']-df_int['YearRemodAdd']
df_int.drop(['Current_year','YearBuilt','YearRemodAdd'],axis = 1, inplace = True)

df_int_final = pd.concat([df_int,mosold,yrsold],axis = 1)


In [ ]:
df_int_final.describe()

In [ ]:
# performing the same function on Object data type
df_obj.info()

In [ ]:
# delelting the column number 2,32,38,49 similer like in train data set
df_obj.drop(['Fence','MiscFeature','FireplaceQu','Alley'],axis = 1,inplace = True)

In [ ]:
df_obj.info()

In [ ]:
bin_enc = BinaryEncoder(cols = ['MSZoning'])
a = bin_enc.fit_transform(df_obj['MSZoning'])

lab_enc = LabelEncoder()
Lot= pd.DataFrame(lab_enc.fit_transform(df_obj['Street']))
Lot.columns=(['Street'])

Lots= pd.DataFrame(lab_enc.fit_transform(df_obj['LotShape']))
Lots.columns=(['LotShape'])

Lotc= pd.DataFrame(lab_enc.fit_transform(df_obj['LotConfig']))
Lotc.columns=(['LotConfig'])

LotS= pd.DataFrame(lab_enc.fit_transform(df_obj['LandSlope']))
LotS.columns=(['LandSlope'])

bin_enc = BinaryEncoder(cols = ['Neighborhood'])
nd = bin_enc.fit_transform(df_obj['Neighborhood'])

bin_enc = BinaryEncoder(cols = ['Condition1'])
cd = bin_enc.fit_transform(df_obj['Condition1'])

bin_enc = BinaryEncoder(cols = ['Condition2'])
cd2 = bin_enc.fit_transform(df_obj['Condition2'])

Bld= pd.DataFrame(lab_enc.fit_transform(df_obj['BldgType']))
Bld.columns=(['BldgType'])

Hst= pd.DataFrame(lab_enc.fit_transform(df_obj['HouseStyle']))
Hst.columns=(['HouseStyle'])

rft= pd.DataFrame(lab_enc.fit_transform(df_obj['RoofStyle']))
rft.columns=(['RoofStyle'])

rftm= pd.DataFrame(lab_enc.fit_transform(df_obj['RoofMatl']))
rftm.columns=(['RoofMatl'])

bin_enc = BinaryEncoder(cols = ['Exterior1st'])
ext1 = bin_enc.fit_transform(df_obj['Exterior1st'])

bin_enc = BinaryEncoder(cols = ['Exterior2nd'])
ext2 = bin_enc.fit_transform(df_obj['Exterior2nd'])

Mas= pd.DataFrame(lab_enc.fit_transform(df_obj['MasVnrType']))
Mas.columns=(['MasVnrType'])

from sklearn.preprocessing import OrdinalEncoder
dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_obj['ExterQual'] = df_obj.ExterQual.map(dic)

dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_obj['ExterCond'] = df_obj.ExterCond.map(dic)

fond= pd.DataFrame(lab_enc.fit_transform(df_obj['Foundation']))
fond.columns=(['Foundation'])

dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_obj['BsmtQual'] = df_obj.BsmtQual.map(dic)

dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_obj['BsmtCond'] = df_obj.BsmtCond.map(dic)

dic = {'AllPub':4,'NoSewr':3,'NoSeWa':2,'ELO':1}
df_obj['Utilities'] = df_obj.Utilities.map(dic)

dic = {'Gd':5,'Av':4,'Mn':3,'No':2,'NA':1}
df_obj['BsmtExposure'] = df_obj.BsmtExposure.map(dic)

dic = {'GLQ':7,'ALQ':6,'BLQ':5,'Rec':4,'LwQ':3,'Unf':2,'NA':1}
df_obj['BsmtFinType1'] = df_obj.BsmtFinType1.map(dic)

dic = {'GLQ':7,'ALQ':6,'BLQ':5,'Rec':4,'LwQ':3,'Unf':2,'NA':1}
df_obj['BsmtFinType2'] = df_obj.BsmtFinType2.map(dic)

dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_obj['HeatingQC'] = df_obj.HeatingQC.map(dic)

heat= pd.DataFrame(lab_enc.fit_transform(df_obj['Heating']))
heat.columns=(['Heating'])

centair= pd.DataFrame(lab_enc.fit_transform(df_obj['CentralAir']))
centair.columns=(['CentralAir'])

elc= pd.DataFrame(lab_enc.fit_transform(df_obj['Electrical']))
elc.columns=(['Electrical'])

dic = {'Ex':5,'Gd':4,'TA':3,'Fa':2,'Po':1}
df_obj['KitchenQual'] = df_obj.KitchenQual.map(dic)

dic = {'Typ':8,'Min1':7,' Min2':6,'Mod':5,'Maj1':4,'Maj2':3,'Sev':2,'Sal':1}
df_obj['Functional'] = df_obj.Functional.map(dic)

dic = {'2Types':7,'Attchd':6,'Basment':5,'BuiltIn':4,'CarPort':3,'Detchd':2,'NA':1}
df_obj['GarageType'] = df_obj.GarageType.map(dic)

dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_obj['GarageQual'] = df_obj.GarageQual.map(dic)

dic = {'Fin':4,'RFn':3,'Unf':2,'NA':1}
df_obj['GarageFinish'] = df_obj.GarageFinish.map(dic)

dic = {'Ex':6,'Gd':5,'TA':4,'Fa':3,'Po':2,'NA':1}
df_obj['GarageCond'] = df_obj.GarageCond.map(dic)

Pav= pd.DataFrame(lab_enc.fit_transform(df_obj['PavedDrive']))
Pav.columns=(['PavedDrive'])

L_C= pd.DataFrame(lab_enc.fit_transform(df_obj['LandContour']))
L_C.columns=(['LandContour'])

bin_enc = BinaryEncoder(cols = ['SaleType'])
S_L_T = bin_enc.fit_transform(df_obj['SaleType'])

bin_enc = BinaryEncoder(cols = ['SaleCondition'])
S_L_C = bin_enc.fit_transform(df_obj['SaleCondition'])

df_obj.drop(['MasVnrType','Exterior2nd','Exterior1st','RoofMatl','RoofStyle','HouseStyle','BldgType',
             'Condition2','Condition1','Neighborhood','LandSlope','LotConfig','LotShape',
             'Street','MSZoning','Foundation','Heating','CentralAir','Electrical','PavedDrive',
             'LandContour','SaleType','SaleCondition'],axis =1,inplace = True)

df_Obj_final = pd.concat([df_obj,S_L_C,S_L_T,L_C,Pav,elc,centair,
                          heat,fond,Mas,ext2,ext1,rftm,rft,Hst,Bld,
                          cd2,cd,nd,LotS,Lotc,Lots,Lot,a],axis =1)



In [ ]:
df_Obj_final.info()

In [110]:
data_Test = pd.concat([df_int_final,df_Obj_final],axis = 1)

In [ ]:
data_Test.info()

In [112]:
Data_Test = data_Test.dropna(axis = 0)

In [ ]:
Data_Test.info()

In [179]:
for i in Data_Test:
  Data_Test['Z_score'] = zscore(Data_Test[i]) 

<ipython-input-179-3d6fe581f398>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Data_Test['Z_score'] = zscore(Data_Test[i])


In [180]:
Data_Test.Z_score

0      0.451243
1      0.451243
2      0.451243
3      0.451243
4      0.451243
         ...   
287    0.451243
288    0.451243
289    0.451243
290   -2.216103
291   -2.216103
Name: Z_score, Length: 266, dtype: float64

In [182]:
Data_Test[Data_Test['Z_score']>3]

Empty DataFrame
Columns: [Id, MSSubClass, LotArea, OverallCond, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, TotRmsAbvGrd, Fireplaces, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, Years_old, remodel_duration, MoSold_0, MoSold_1, MoSold_2, MoSold_3, YrSold_0, YrSold_1, YrSold_2, Utilities, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, Functional, GarageType, GarageFinish, GarageQual, GarageCond, SaleCondition_0, SaleCondition_1, SaleCondition_2, SaleType_0, SaleType_1, SaleType_2, LandContour, PavedDrive, Electrical, CentralAir, Heating, Foundation, MasVnrType, Exterior2nd_0, Exterior2nd_1, Exterior2nd_2, Exterior2nd_3, Exterior1st_0, Exterior1st_1, Exterior1st_2, Exterior1st_3, RoofMatl, RoofStyle, HouseStyle, BldgType, Condition2_0, Condition2_1, Condition1_0, Condition1_1, Condition1_2, Condition1_3, Neighborhood_0, Neighborhood_1, Neighborhood_2, Neighborhood_3, Neighborhood_4, LandSlope, LotConfig, LotShape, Street, MSZoning_0, MSZoning_1, MSZoning_2, Z_score]
Index: []

[0 rows x 97 columns]

In [183]:
Data_Test[Data_Test['Z_score']<-3]

Empty DataFrame
Columns: [Id, MSSubClass, LotArea, OverallCond, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, TotRmsAbvGrd, Fireplaces, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, Years_old, remodel_duration, MoSold_0, MoSold_1, MoSold_2, MoSold_3, YrSold_0, YrSold_1, YrSold_2, Utilities, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, Functional, GarageType, GarageFinish, GarageQual, GarageCond, SaleCondition_0, SaleCondition_1, SaleCondition_2, SaleType_0, SaleType_1, SaleType_2, LandContour, PavedDrive, Electrical, CentralAir, Heating, Foundation, MasVnrType, Exterior2nd_0, Exterior2nd_1, Exterior2nd_2, Exterior2nd_3, Exterior1st_0, Exterior1st_1, Exterior1st_2, Exterior1st_3, RoofMatl, RoofStyle, HouseStyle, BldgType, Condition2_0, Condition2_1, Condition1_0, Condition1_1, Condition1_2, Condition1_3, Neighborhood_0, Neighborhood_1, Neighborhood_2, Neighborhood_3, Neighborhood_4, LandSlope, LotConfig, LotShape, Street, MSZoning_0, MSZoning_1, MSZoning_2, Z_score]
Index: []

[0 rows x 97 columns]

No outliers 

In [184]:
# Now handling MultiCollinearity 
from statsmodels.stats.outliers_influence import variance_inflation_factor
b = pd.DataFrame()
b['Vif'] = Data_Test.columns
b["Vif"] = [variance_inflation_factor(Data_Test.values, i)
                          for i in range(len(Data_Test.columns))]
b['Vif'].head(20)

/usr/local/lib/python3.8/dist-packages/statsmodels/stats/outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)
/usr/local/lib/python3.8/dist-packages/statsmodels/regression/linear_model.py:1715: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - self.ssr/self.centered_tss


0      1.390048
1     20.478404
2      3.337660
3      4.016998
4           inf
5           inf
6           inf
7           inf
8           inf
9           inf
10          inf
11          inf
12     3.617860
13     1.736733
14     4.852792
15     4.584491
16     3.472682
17     2.633451
18     7.791321
19     2.376271
Name: Vif, dtype: float64

In [185]:
Df_Tr = Data_Train.drop(Data_Train.columns[[4,5,6,7,8,9,10,11]],axis = 1)

In [187]:
Data_Test.drop(['Z_score'],axis = 1,inplace = True)

/usr/local/lib/python3.8/dist-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [188]:
c= pd.DataFrame()
c['vif'] = Df_Tr.columns

c["Vif"] = [variance_inflation_factor(Df_Tr.values, i)
                          for i in range(len(Df_Tr.columns))]

/usr/local/lib/python3.8/dist-packages/statsmodels/regression/linear_model.py:1715: RuntimeWarning: divide by zero encountered in double_scalars
  return 1 - self.ssr/self.centered_tss
/usr/local/lib/python3.8/dist-packages/statsmodels/stats/outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


In [189]:
c[c['Vif']>9]

vif        Vif
48  SaleCondition_2  10.130705
51       SaleType_2  10.530818
91       MSZoning_2        inf
92          Z_score        inf

In [190]:
Df_Test = Df_Tr.drop(Df_Tr.columns[[48,51,91,62]],axis = 1)

In [191]:
Df_Test

Id  MSSubClass  LotArea  OverallCond  BsmtFullBath  BsmtHalfBath  \
0      127         120     4928            5             0             0   
1      889          20    15865            6             1             0   
2      793          60     9920            5             1             0   
3      110          20    11751            6             0             0   
4      422          20    16635            7             0             1   
...    ...         ...      ...          ...           ...           ...   
1162  1381          30     8212            3             1             0   
1163   289          20     9819            5             0             0   
1165   196         160     2280            6             0             0   
1166    31          70     8500            4             0             0   
1167   617          60     7861            5             1             0   

      FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  ...  Neighborhood_2  \
0            2         0             2             1  ...               0   
1            2         0             4             1  ...               0   
2            2         1             3             1  ...               0   
3            2         0             3             1  ...               1   
4            2         0             3             1  ...               1   
...        ...       ...           ...           ...  ...             ...   
1162         1         0             2             1  ...               1   
1163         1         0             3             1  ...               1   
1165         2         1             3             1  ...               0   
1166         1         0             3             1  ...               0   
1167         2         1             3             1  ...               1   

      Neighborhood_3  Neighborhood_4  LandSlope  LotConfig  LotShape  Street  \
0                  0               1          0          4         0       1   
1                  1               0          1          4         0       1   
2                  1               1          0          1         0       1   
3                  0               0          0          4         0       1   
4                  0               0          0          2         0       1   
...              ...             ...        ...        ...       ...     ...   
1162               1               1          0          4         3       1   
1163               1               0          0          4         0       1   
1165               0               1          0          2         3       1   
1166               0               0          0          4         3       1   
1167               0               1          0          4         0       1   

      MSZoning_0  MSZoning_1   Z_score  
0              0           0  0.411189  
1              0           0  0.411189  
2              0           0  0.411189  
3              0           0  0.411189  
4              0           0  0.411189  
...          ...         ...       ...  
1162           0           0  0.411189  
1163           0           0  0.411189  
1165           0           0  0.411189  
1166           1           0  0.411189  
1167           0           0  0.411189  

[1051 rows x 89 columns]

We have done the outliers check on the test data set . we have checked the multicolinearity on test data set also 

In [192]:
 # create x = freature and y = label
 x = DF.drop(['SalePrice'],axis = 1)
 y = DF['SalePrice']

In [195]:
a = Df_Test.drop(['SalePrice'],axis = 1)
b = Df_Test['SalePrice']

In [193]:
from sklearn.linear_model import LinearRegression

lm = LinearRegression()
lm.fit(x,y)

LinearRegression()

In [194]:
# print intercept and coefficients
print(lm.intercept_) # this is c or b interception 
print(lm.coef_) # this is M (slop)

-417590.04315524467
[-9.10530732e-01 -1.37359370e+02  4.30044369e-01  3.93900515e+03
  6.72586838e+03 -1.07384838e+03  2.36691468e+04  1.61416206e+04
 -3.20074194e+03 -2.41015211e+04  1.12109357e+04  1.25670003e+04
  1.09298498e+04  4.19479181e+01  2.30163253e+01  2.48828218e+01
  2.08969154e+01  7.68582885e+01  3.88063493e+01  2.51635708e+01
  7.60961671e-01  8.74484028e+01  9.93996126e+01  1.80707410e+03
 -6.40566240e+02 -1.46126184e+03  1.05635852e+03 -3.30382136e+03
 -1.86864841e+03 -2.69344025e+03  1.52652159e+04 -6.07087797e+03
  1.73348522e+04 -1.78432900e+03  6.83937055e+03  3.10477475e+03
 -6.49381621e+02  1.14995348e+03  1.53837202e+04  6.13917461e+03
  3.37365467e+02  1.57128687e+03  4.35919355e+03 -5.74595053e+03
  6.25375178e+02  5.68030975e+03  9.40776792e+03  6.29901530e+03
  2.35478430e+03  4.17820866e+03  1.92408070e+03 -8.83298598e+02
  7.46275914e+03  2.40791007e+03  2.85580218e+03  2.12566950e+03
  2.78595946e+03 -6.15387929e+03  9.24301421e+03 -1.36607017e+03
  4.8

In [196]:
predicted_sales = lm.predict(a)

/usr/local/lib/python3.8/dist-packages/sklearn/base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- Utilities
- Z_score
Feature names seen at fit time, yet now missing:
- Exterior2nd_2
- MSZoning_2

  warnings.warn(message, FutureWarning)


In [198]:
print(predicted_sales)

[115902.73290245 285521.41991032 213997.03046475 ... 138970.72021893
  41700.34474683 204726.73705099]


In [199]:
from sklearn.metrics import r2_score
r2_score(y_true=b,y_pred=predicted_sales)

0.6625896828665749